src/d3a/models/strategy/__init__.py

```diff
@@ -402,13 +402,25 @@ def posted_bid_energy(self, market_id):
             return 0.0
         return sum(b.energy for b in self._bids[market_id])
 
-    def remove_bid_from_pending(self, bid_id, market_id):
+    def remove_bid_from_pending(self, market_id, bid_id=None):
         market = self.area.get_future_market_from_id(market_id)
         if market is None:
             return
-        if bid_id in market.bids.keys():
-            market.delete_bid(bid_id)
-        self._bids[market.id] = [bid for bid in self._bids[market.id] if bid.id != bid_id]
+        if bid_id is None:
+            # delete all bids associated with this strategy
+            deleted_bids_dict = {b_id: bid.energy for b_id, bid in market.bids.items()
+                                 if bid.buyer == self.owner.name}
```

Comment:
I know that this has a similar effect, but I would very much prefer to use the bid cache here and not access the market (self._bids are the bids posted by this strategy on a market). Second, the energy should not be needed as value of the dictionary.

Reply:
I agree with you that using the `_bids` member for accessing energy (sorry, didnt think).
But we need the energy in the external storage to account for the already offered energy: https://github.com/gridsingularity/d3a/blob/af70782338868c2fb978367c6e820ae4c9e447ae/src/d3a/models/strategy/external_strategies/storage.py#L181

Reply:
I see now why the energy is required to be returned. In that case, I would prefer to make this calculation a bit more stateless, by exposing another function called `total_bid_energy` , which would automatically calculate the `self.state.offered_buy_kWh` by summing up the energy of all posted bids (`_bids`). That way you will not have to keep track of the state by decrementing the energy of the deleted bid, but it can be recalculated every time a bid is posted or deleted. If it is not that clear let's discuss offline

Reply:
Ok, I like that idea! Didnt like the extraction of energy in the `remove_bid_from_pending` function. 
So, will proceed your proposed way. I guess that this `total_bid_energy ` could live in the BidEnabledStrategy class right?

Reply:
Exactly, it would be great to be a small function in this class, thanks

Reply:
OMG! there is already such function: `posted_bid_energy` Sorry!